niff-1.02/niffsdk1.02/niffio/doc/man/register.man
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH register 3 "niffio" "14/Jun/96" "NIFFIO Documentation"
.IX Title "register 3"
.UC
.IX Name "register - parser callback registration routines"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
register \- parser callback registration routines
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterDefaultList()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterDefaultChunk()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterDefaultAtomicChunk()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterDefaultTag()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterForm()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterList()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterAtomicChunk()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterListXXX()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterChunkXXX()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fINIFFIORegisterTagXXX()\fR
.SH "NIFFIORegisterDefaultList"
.IX Header "NIFFIORegisterDefaultList"
Register default list callbacks
.PP
.Vb 4
\&     RIFFIOSuccess
\&     NIFFIORegisterDefaultList(NIFFIOParser *pparser,
\&                               NIFFIOChunkCallback cbStart,
\&                               NIFFIOChunkCallback cbEnd)
.Ve
.SH "NIFFIORegisterDefaultTaggedChunk"
.IX Header "NIFFIORegisterDefaultTaggedChunk"
Register default chunk callbacks
.PP
.Vb 4
\&     RIFFIOSuccess
\&     NIFFIORegisterDefaultTaggedChunk(NIFFIOParser *pparser,
\&                                      NIFFIOChunkCallback cbStart,
\&                                      NIFFIOChunkCallback cbEnd)
.Ve
.SH "NIFFIORegisterDefaultAtomicChunk"
.IX Header "NIFFIORegisterDefaultAtomicChunk"
Register a default callback for Atomic chunks
.PP
.Vb 3
\&     RIFFIOSuccess
\&     NIFFIORegisterDefaultAtomicChunk(NIFFIOParser *pparser,
\&                                      NIFFIOChunkCallback cb)
.Ve
.SH "NIFFIORegisterDefaultTag"
.IX Header "NIFFIORegisterDefaultTag"
Register a default callback for tags
.PP
.Vb 3
\&     RIFFIOSuccess
\&     NIFFIORegisterDefaultTag(NIFFIOParser *pparser,
\&                              NIFFIOTagCallback cbTag)
.Ve
.SH "NIFFIORegisterForm"
.IX Header "NIFFIORegisterForm"
Register callbacks for a Niff form
.PP
.Vb 4
\&     RIFFIOSuccess
\&     NIFFIORegisterForm(NIFFIOParser *pparser,
\&                        NIFFIOChunkCallback cbStart,
\&                        NIFFIOChunkCallback cbEnd)
.Ve
.SH "NIFFIORegisterList"
.IX Header "NIFFIORegisterList"
Register callbacks for specific lists
.PP
.Vb 5
\&     RIFFIOSuccess
\&     NIFFIORegisterList(NIFFIOParser *pparser,
\&                        FOURCC fccType,
\&                        NIFFIOChunkCallback cbStart,
\&                        NIFFIOChunkCallback cbEnd)
.Ve
.SH "NIFFIORegisterAtomicChunk"
.IX Header "NIFFIORegisterAtomicChunk"
Register a callback for specific Atomic chunks
.PP
.Vb 4
\&     RIFFIOSuccess
\&     NIFFIORegisterAtomicChunk(NIFFIOParser *pparser,
\&                               FOURCC fccId,
\&                               NIFFIOChunkCallback cb)
.Ve
Registration functions for Lists
.PP
.Vb 8
\&     #define REG_LIST(nifftype)                                                   \e
\&     RIFFIOSuccess                                                                \e
\&     NIFFIORegisterList##nifftype(NIFFIOParser *pparser,                          \e
\&                                  NIFFIOChunkCallback cbStart,                    \e
\&                                  NIFFIOChunkCallback cbEnd)                      \e
\&     {                                                                            \e
\&          return NIFFIORegisterList(pparser, nifflist##nifftype, cbStart, cbEnd); \e
\&     }
.Ve
Registration functions for atomic, raw chunks
.PP
.Vb 9
\&     #define REG_ATOMIC(nifftype)                                            \e
\&     RIFFIOSuccess                                                           \e
\&     NIFFIORegisterChunk##nifftype(                                          \e
\&                                   NIFFIOParser *pparser,                    \e
\&                                   NIFFIOChunkCallback cb)                   \e
\&     {                                                                       \e
\&          return NIFFIORegisterAtomicChunk(pparser, niffckid##nifftype, cb); \e
\&     }
\&     
.Ve
Registration functions for tagable, raw Chunks
.PP
.Vb 13
\&     
\&     #define REG_TAGRAW(nifftype)                                    \e

niff-1.02/niffsdk1.02/niffio/src/lib/niffio/register.c
#ifndef lint
static char rcsid[] = 
"$Id: register.c,v 1.3 1996/06/15 02:53:54 tim Exp $";
#endif
/*
 * Public Domain 1995,1996 Timothy Butler
 *
 * THIS DOCUMENT IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 */
/***************************************************************************/
/*
 * NAME
 * ====
 * register - parser callback registration routines
 *
 * SYNOPSIS
 * ======== 
 *
 * - NIFFIORegisterDefaultList()
 * - NIFFIORegisterDefaultChunk()
 * - NIFFIORegisterDefaultAtomicChunk()
 * - NIFFIORegisterDefaultTag()
 *
 * - NIFFIORegisterForm()
 * - NIFFIORegisterList()
 * - NIFFIORegisterAtomicChunk()
 * 
 * - NIFFIORegisterListXXX()
 * - NIFFIORegisterChunkXXX()
 * - NIFFIORegisterTagXXX()
 */
/***************************************************************************/

#include <assert.h>
#include <stdlib.h>

#include "niffio.h"
#include "niffiop.h"

/***************************************************************************/
/*
 * NIFFIORegisterDefaultList
 * =========================
 * Register default list callbacks
 */
RIFFIOSuccess
NIFFIORegisterDefaultList(NIFFIOParser *pparser,
                          NIFFIOChunkCallback cbStart,
                          NIFFIOChunkCallback cbEnd)
/***************************************************************************/
{
    assert (pparser != 0);

    pparser->defaultList.cbStart = cbStart;
    pparser->defaultList.cbEnd = cbEnd;

    return RIFFIO_OK;
}

/***************************************************************************/
/*
 * NIFFIORegisterDefaultTaggedChunk
 * ================================
 * Register default chunk callbacks
 */
RIFFIOSuccess
NIFFIORegisterDefaultTaggedChunk(NIFFIOParser *pparser,
                                 NIFFIOChunkCallback cbStart,
                                 NIFFIOChunkCallback cbEnd)
/***************************************************************************/
{
    assert (pparser != 0);

    pparser->defaultTaggedChunk.cbStart = cbStart;
    pparser->defaultTaggedChunk.cbEnd = cbEnd;

    return RIFFIO_OK;
}

/***************************************************************************/
/*
 * NIFFIORegisterDefaultAtomicChunk
 * ================================
 * Register a default callback for Atomic chunks
 */
RIFFIOSuccess
NIFFIORegisterDefaultAtomicChunk(NIFFIOParser *pparser,
                                 NIFFIOChunkCallback cb)
/***************************************************************************/
{
    assert (pparser != 0);
        
    pparser->defaultAtomicChunk.cb = cb;

    return RIFFIO_OK;
}

/***************************************************************************/
/*
 * NIFFIORegisterDefaultTag
 * ========================
 * Register a default callback for tags
 */
RIFFIOSuccess
NIFFIORegisterDefaultTag(NIFFIOParser *pparser,
                         NIFFIOTagCallback cbTag)
/***************************************************************************/
{
    assert (pparser != 0);

    pparser->defaultTag.cbTag = cbTag;

    return RIFFIO_OK;
        
}

/***************************************************************************/
/*
 * NIFFIORegisterForm
 * ==================
 * Register callbacks for a Niff form
 */
RIFFIOSuccess
NIFFIORegisterForm(NIFFIOParser *pparser,
                   NIFFIOChunkCallback cbStart,
                   NIFFIOChunkCallback cbEnd)
/***************************************************************************/
{
    assert (pparser != 0);
        
    pparser->formentry.cbStart = cbStart;
    pparser->formentry.cbEnd = cbEnd;

    return RIFFIO_OK;

}

/***************************************************************************/
/*
 * NIFFIORegisterList
 * ==================
 * Register callbacks for specific lists
 */
RIFFIOSuccess
NIFFIORegisterList(NIFFIOParser *pparser,
                   FOURCC fccType,
                   NIFFIOChunkCallback cbStart,
                   NIFFIOChunkCallback cbEnd)
/***************************************************************************/
{
        
    char strModule[] = "NIFFIORegisterList";
    char strType[RIFFIO_FOURCC_LIM];

    NIFFIOPListEntry *pListEntry; /* newly allocated entry */
    RIFFIOFCCTableEntry tblEntry; /* interface to RIFFIOFCCTableMakeEntry */

    assert (pparser != 0);

    RIFFIOFOURCCToString(fccType, strType);
        
    if (!RIFFIOFOURCCIsValid(fccType))
    {
        RIFFIOError(strModule, "Invalid FOURCC type <%s>", strType);
        return RIFFIO_FAIL;
    }

    /*
     * Allocate a new list entry structure to hold our callbacks
     */
    pListEntry = (NIFFIOPListEntry *) malloc(sizeof(NIFFIOPListEntry));
    if (pListEntry == 0)
    {
        RIFFIOError(strModule, 
                    "Failed to allocate list callback entry, type <%s>",
                    strType);

        return RIFFIO_FAIL;
    }

    /*
     * Initialize the new entry
     */
    pListEntry->cbStart = cbStart;
    pListEntry->cbEnd = cbEnd;

    /*
     * Make the entry in the list callback table
     */
    tblEntry.fcc = fccType;
    tblEntry.value.pv = pListEntry;

    if (! RIFFIOFCCTableMakeEntry(pparser->pfcctblLists, tblEntry))
    {
        /*
         * Clean up on failure
         */
        free(pListEntry);

        RIFFIOError(strModule,
                    "Failed to install new list callback entry, type <%s>",
                    strType);
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;

}

/***************************************************************************/
/*
 * NIFFIORegisterAtomicChunk
 * =========================
 * Register a callback for specific Atomic chunks
 */
RIFFIOSuccess
NIFFIORegisterAtomicChunk(NIFFIOParser *pparser,
                          FOURCC fccId,
                          NIFFIOChunkCallback cb)
/***************************************************************************/
{

    char strModule[] = "NIFFIORegisterAtomicChunk";
    char strId[RIFFIO_FOURCC_LIM];

    NIFFIOPAtomicChunkEntry *pAtomicChunkEntry;   /* newly allocated entry */
    RIFFIOFCCTableEntry tblEntry;  /* Interface to RIFFIOFCCTableMakeEntry */

    assert (pparser != 0);

    RIFFIOFOURCCToString(fccId, strId);

    if (! RIFFIOFOURCCIsValid(fccId))
    {
        RIFFIOError(strModule, "Invalid FOURCC id <%s>", strId);
        return RIFFIO_FAIL;
    }

    /*
     * Allocate a new list entry structure to hold our callbacks
     */
    pAtomicChunkEntry = 
        (NIFFIOPAtomicChunkEntry *) malloc(sizeof(NIFFIOPAtomicChunkEntry));
    if (pAtomicChunkEntry == 0)
    {
        RIFFIOError(strModule, 
                    "Failed to allocate chunk callback entry, id <%s>",
                    strId);
        return RIFFIO_FAIL;
    }

    /*
     * Initialize the new entry
     */
    pAtomicChunkEntry->cb = cb;

    /*
     * Make the entry in the list callback table
     */
    tblEntry.fcc = fccId;
    tblEntry.value.pv = pAtomicChunkEntry;

    if (! RIFFIOFCCTableMakeEntry(pparser->pfcctblAtomicChunks, tblEntry))
    {
        /*
         * Clean up on failure
         */
        free(pAtomicChunkEntry);

        RIFFIOError(strModule,
                    "Failed to install new chunk callback entry, id <%s>",
                    strId);
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;

}

/*
 * NIFFIOPRegisterTaggedCookedChunk
 * ================================
 * (Private)
 * Register callbacks, including a dispatching routine
 * for specific composite chunks
 */
RIFFIOSuccess
NIFFIOPRegisterTaggedCookedChunk(NIFFIOParser *pparser,
                                 FOURCC fccId,
                                 NIFFIOPReader cbRead,
                                 NIFFIOCookedChunkCallback cbStart,
                                 NIFFIOCookedChunkCallback cbEnd)
{
    
    char strModule[] = "NIFFIOPRegisterTaggedCookedChunk";
    char strId[RIFFIO_FOURCC_LIM];
    
    NIFFIOPTaggedChunkEntry *pTaggedChunkEntry; /* a new value for our table */
    
    RIFFIOFCCTableEntry tblEntry;    /* interface to NIFFIOFCCTableMakeEntry */
    
    assert (pparser != 0);
    
    RIFFIOFOURCCToString(fccId, strId);
    
    if (! RIFFIOFOURCCIsValid(fccId))
    {
        RIFFIOError(strModule, "Invalid FOURCC id <%s>", strId);
        return RIFFIO_FAIL;
    }

    /*
     * Allocate a new list entry structure to hold our callbacks
     */
    pTaggedChunkEntry = (NIFFIOPTaggedChunkEntry *) 
        malloc(sizeof(NIFFIOPTaggedChunkEntry));
        
    if (! pTaggedChunkEntry)
    {
        RIFFIOError(strModule, 
                    "Failed to allocate chunk callback entry, id <%s>",
                    strId);

        return RIFFIO_FAIL;
    }

    /*
     * Initialize the new entry 
     */
    pTaggedChunkEntry->cbRead = cbRead;
    pTaggedChunkEntry->cbStart.cooked = cbStart;
    pTaggedChunkEntry->cbEnd.cooked = cbEnd;

    /*
     * Make the entry in the callback table
     */
    tblEntry.fcc = fccId;
    tblEntry.value.pv = pTaggedChunkEntry;

    if (! RIFFIOFCCTableMakeEntry(pparser->pfcctblTaggedChunks, tblEntry))
    {
        /*
         * Clean up on failure 
         */
        free(pTaggedChunkEntry);

        RIFFIOError(strModule,
                    "Failed to install new chunk callback entry, id <%s>",
                    strId);
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;

}


/*
 * NIFFIOPRegisterTaggedRawChunk
 * =============================
 * (Private)
 * Register callbacks, for specific tagged, raw chunks
 * These are taggable chunks with an empty fixed portion.
 * (CLT entry == 0)
 */
RIFFIOSuccess
NIFFIOPRegisterTaggedRawChunk(NIFFIOParser *pparser,
                              FOURCC fccId,
                              NIFFIOChunkCallback cbStart,
                              NIFFIOChunkCallback cbEnd)
{

    char strModule[] = "NIFFIOPRegisterTaggedRawChunk";
    char strId[RIFFIO_FOURCC_LIM];

    NIFFIOPTaggedChunkEntry *pTaggedChunkEntry; /* a new value for our table */

    RIFFIOFCCTableEntry tblEntry;    /* interface to NIFFIOFCCTableMakeEntry */

    assert (pparser != 0);

    RIFFIOFOURCCToString(fccId, strId);

    if (! RIFFIOFOURCCIsValid(fccId))
    {
        RIFFIOError(strModule, "Invalid FOURCC id <%s>", strId);
        return RIFFIO_FAIL;
    }

    /*
     * Allocate a new list entry structure to hold our callbacks
     */
    pTaggedChunkEntry = (NIFFIOPTaggedChunkEntry *) 
        malloc(sizeof(NIFFIOPTaggedChunkEntry));
        
    if (! pTaggedChunkEntry)
    {
        RIFFIOError(strModule, 
                    "Failed to allocate chunk callback entry, id <%s>",
                    strId);

        return RIFFIO_FAIL;
    }

    /*
     * Initialize the new entry 
     */
    pTaggedChunkEntry->cbRead = 0;
    pTaggedChunkEntry->cbStart.raw = cbStart;
    pTaggedChunkEntry->cbEnd.raw = cbEnd;

    /*
     * Make the entry in the callback table
     */
    tblEntry.fcc = fccId;
    tblEntry.value.pv = pTaggedChunkEntry;

    if (! RIFFIOFCCTableMakeEntry(pparser->pfcctblTaggedChunks, tblEntry))
    {
        /*
         * Clean up on failure 
         */
        free(pTaggedChunkEntry);

        RIFFIOError(strModule,
                    "Failed to install new chunk callback entry, id <%s>",
                    strId);
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;

}

/*
 * NIFFIOPRegisterCookedTag
 * ========================
 * (private)
 * Register a callback and a reader for a specific tag/FOURCC combination
 */ 
RIFFIOSuccess
NIFFIOPRegisterCookedTag(NIFFIOParser *pparser,
                         BYTE tagid,
                         FOURCC fcc,
                         NIFFIOPReader cbRead,
                         NIFFIOCookedTagCallback cbTag)
{
    NIFFIOPTagEntry tagentry;

    assert (pparser != 0);
    assert (pparser->ptcbtblTags != 0);
    assert (cbTag != 0);

    tagentry.fcc = fcc;
    tagentry.cbRead = cbRead;
    tagentry.cbTag.cooked = cbTag;
        
    return NIFFIOPTCBTableMakeEntry(pparser->ptcbtblTags, tagid, &tagentry);
        
}

/*
 * NIFFIOPRegisterRawTag
 * =====================
 * (private)
 * Register a callback for a specific tag/FOURCC combination
 */ 
RIFFIOSuccess
NIFFIOPRegisterRawTag(NIFFIOParser *pparser,
                      BYTE tagid,
                      FOURCC fcc,
                      NIFFIOTagCallback cbTag)
{
    NIFFIOPTagEntry tagentry;

    assert (pparser != 0);
    assert (pparser->ptcbtblTags != 0);
    assert (cbTag != 0);

    tagentry.fcc = fcc;
    tagentry.cbRead = 0;
    tagentry.cbTag.raw = cbTag;
        
    return NIFFIOPTCBTableMakeEntry(pparser->ptcbtblTags, tagid, &tagentry);
        

}

/***************************************************************************/
/*
 * Registration functions for Lists
 */
#define REG_LIST(nifftype)                                                   \
RIFFIOSuccess                                                                \
NIFFIORegisterList##nifftype(NIFFIOParser *pparser,                          \
                             NIFFIOChunkCallback cbStart,                    \
                             NIFFIOChunkCallback cbEnd)                      \
{                                                                            \
     return NIFFIORegisterList(pparser, nifflist##nifftype, cbStart, cbEnd); \
}
/***************************************************************************/

REG_LIST(SetupSection)
REG_LIST(Parts)
REG_LIST(RiffInfo)
REG_LIST(Groupings)
REG_LIST(FontDescs)
REG_LIST(CustomGraphics)

REG_LIST(DataSection)
REG_LIST(Page)
REG_LIST(System)
REG_LIST(Staff)
                                                         
#undef REG_LIST

/***************************************************************************/
/*
 * Registration functions for atomic, raw chunks
 */
#define REG_ATOMIC(nifftype)                                            \
RIFFIOSuccess                                                           \
NIFFIORegisterChunk##nifftype(                                          \
                              NIFFIOParser *pparser,                    \
                              NIFFIOChunkCallback cb)                   \
{                                                                       \
     return NIFFIORegisterAtomicChunk(pparser, niffckid##nifftype, cb); \
}

/***************************************************************************/

REG_ATOMIC(ChnkLenTable)
REG_ATOMIC(EpsGraphic)
REG_ATOMIC(PsType1Font)
REG_ATOMIC(PsType3Font)
REG_ATOMIC(StringTable)

#undef REG_ATOMIC

/***************************************************************************/
/*
 * Registration functions for tagable, raw Chunks
 */

#define REG_TAGRAW(nifftype)                                    \
RIFFIOSuccess                                                   \
NIFFIORegisterChunk##nifftype(                                  \
                              NIFFIOParser *pparser,            \
                              NIFFIOChunkCallback cbStart,      \
                              NIFFIOChunkCallback cbEnd)        \
{                                                               \
     return NIFFIOPRegisterTaggedRawChunk(pparser,              \
                                          niffckid##nifftype,   \
                                          cbStart,              \
                                          cbEnd);               \
}
/***************************************************************************/

REG_TAGRAW(PageHeader)
REG_TAGRAW(StaffHeader)
REG_TAGRAW(SystemHeader)

REG_TAGRAW(AugDot)
REG_TAGRAW(Glissando)
REG_TAGRAW(Portamento)
REG_TAGRAW(Slur)
REG_TAGRAW(Stem)
REG_TAGRAW(TagActivate)
REG_TAGRAW(TagInactivate)
REG_TAGRAW(Tie)
REG_TAGRAW(Tuplet)

#undef REG_TAGRAW


/***************************************************************************/
/*
 * Registration functions for tagable, cooked Chunks
 */
#define REG_TAGCOOK(nifftype)                                           \
RIFFIOSuccess                                                           \
NIFFIORegisterChunk##nifftype(                                          \
   NIFFIOParser *pparser,                                               \
   RIFFIOSuccess (*cbStart)(NIFFIOChunkContext *, niff##nifftype *),    \
   RIFFIOSuccess (*cbEnd)(NIFFIOChunkContext *, niff##nifftype *))      \
{                                                                       \
   return NIFFIOPRegisterTaggedCookedChunk(                             \
        pparser,                                                        \
        niffckid##nifftype,                                             \
        (NIFFIOPReader) NIFFIOReadniff##nifftype,                       \
        (NIFFIOCookedChunkCallback) cbStart,                            \
        (NIFFIOCookedChunkCallback) cbEnd);                             \
}
/***************************************************************************/

REG_TAGCOOK(DefaultValues)
REG_TAGCOOK(FontDescription)
REG_TAGCOOK(NiffInfo)
REG_TAGCOOK(Part)
REG_TAGCOOK(StaffGrouping)
REG_TAGCOOK(TimeSlice)

REG_TAGCOOK(Accidental)
REG_TAGCOOK(AltEndingGraphic)
REG_TAGCOOK(Arpeggio)
REG_TAGCOOK(Articulation)
REG_TAGCOOK(Barline)
REG_TAGCOOK(Beam)
REG_TAGCOOK(ChordSymbol)
REG_TAGCOOK(Clef)
REG_TAGCOOK(CustomGraphicChk)
REG_TAGCOOK(Dynamic)
REG_TAGCOOK(FiguredBass)
REG_TAGCOOK(Fingering)
REG_TAGCOOK(GuitarGrid)
REG_TAGCOOK(GuitarTabNum)
REG_TAGCOOK(Hairpin)
REG_TAGCOOK(HarpPedal)
REG_TAGCOOK(KeySignature)
REG_TAGCOOK(KeySignNonstandard)
REG_TAGCOOK(Line)
REG_TAGCOOK(Lyric)
REG_TAGCOOK(MeasureNumbering)
REG_TAGCOOK(MidiDataStream)
REG_TAGCOOK(FontSymbol)
REG_TAGCOOK(Notehead)
REG_TAGCOOK(OctaveSign)
REG_TAGCOOK(Ornament)
REG_TAGCOOK(Parenthesis)
REG_TAGCOOK(PedalPiano)
REG_TAGCOOK(PedalOrgan)
REG_TAGCOOK(RehearsalMark)
REG_TAGCOOK(RepeatSign)
REG_TAGCOOK(Rest)
REG_TAGCOOK(SystemSeparation)
REG_TAGCOOK(TempoMarking)
REG_TAGCOOK(TempoMarkNonstandard)
REG_TAGCOOK(Text)
REG_TAGCOOK(TimeSignature)
REG_TAGCOOK(TimeSigNonstandard)
REG_TAGCOOK(Tremolo)

#undef REG_TAGCOOK



/***************************************************************************/
/*
 * Registration functions for raw Tags
 */
#define REG_TAGRAW(nifftype)                                            \
RIFFIOSuccess                                                           \
NIFFIORegisterTag##nifftype(                                            \
                            NIFFIOParser *pparser,                      \
                            FOURCC fcc,                                 \
                            NIFFIOTagCallback cb)                       \
{                                                                       \
     return NIFFIOPRegisterRawTag(pparser, nifftag##nifftype, fcc, cb); \
}

/***************************************************************************/

REG_TAGRAW(EndOfSystem)
REG_TAGRAW(GuitarTabTag)
REG_TAGRAW(Invisible)
REG_TAGRAW(LargeSize)
REG_TAGRAW(MultiNodeEndOfSyst)
REG_TAGRAW(MultiNodeStartOfSyst)
REG_TAGRAW(Silent)
REG_TAGRAW(SlashedStem)
REG_TAGRAW(SmallSize)
REG_TAGRAW(SpacingByPart)
REG_TAGRAW(SplitStem)
REG_TAGRAW(StaffName)

#undef REG_TAGRAW


/***************************************************************************/
/*
 * Registration functions for cooked Tags
 */
#define REG_TAGCOOK(nifftype)                                           \
RIFFIOSuccess                                                           \
NIFFIORegisterTag##nifftype(                                            \
      NIFFIOParser *pparser,                                            \
      FOURCC fcc,                                                       \
      RIFFIOSuccess (*cb)(NIFFIOTagContext *, niff##nifftype *))        \
{                                                                       \
     return NIFFIOPRegisterCookedTag(                                   \
                pparser,                                                \
                nifftag##nifftype,                                      \
                fcc,                                                    \
                (NIFFIOPReader) NIFFIOReadniff##nifftype,               \
                (NIFFIOCookedTagCallback) cb);                          \
}

/***************************************************************************/
   

REG_TAGCOOK(AbsPlacement)
REG_TAGCOOK(AltEnding)
REG_TAGCOOK(AnchorOverride)
REG_TAGCOOK(ArticDirection)
REG_TAGCOOK(BezierIncoming)
REG_TAGCOOK(BezierOutgoing)
REG_TAGCOOK(ChordSymbolsOffset)
REG_TAGCOOK(CustomFontChar)
REG_TAGCOOK(CustomGraphicTag)
REG_TAGCOOK(FannedBeam)
REG_TAGCOOK(FigBassOffset)
REG_TAGCOOK(FontID)
REG_TAGCOOK(GraceNote)
REG_TAGCOOK(GuitarGridOffset)
REG_TAGCOOK(Height)
REG_TAGCOOK(ID)
REG_TAGCOOK(LineQuality)
REG_TAGCOOK(LogicalPlacement)
REG_TAGCOOK(LyricVerseOffset)
REG_TAGCOOK(MidiPerformance)
REG_TAGCOOK(NumberOfFlags)
REG_TAGCOOK(NumberOfNodes)
REG_TAGCOOK(NumStaffLines)
REG_TAGCOOK(Ossia)
REG_TAGCOOK(PartDescOverride)
REG_TAGCOOK(PartID)
REG_TAGCOOK(RefPtOverride)
REG_TAGCOOK(RehearsalOffset)
REG_TAGCOOK(RestNumeral)
REG_TAGCOOK(StaffStep)
REG_TAGCOOK(Thickness)
REG_TAGCOOK(TieDirection)
REG_TAGCOOK(TupletDesc)
REG_TAGCOOK(NumberStyle)
REG_TAGCOOK(VoiceID)
REG_TAGCOOK(Width)


#undef REG_TAGCOOK


